### PR TITLE
Implement GET and PUT for tags config

### PR DESCRIPTION
**Pull request by demyanenko**

### CODE REVIEW

File: server/Controllers/ApiController.cs [hunk 1]
```diff
@@ -146,6 +147,48 @@ public IActionResult PutOne(Guid id, [FromBody]PointJson pointJson)
...
+            {
+                return new BadRequestObjectResult(
+                    new InvalidOperationException(
+                        $"Duplicate tag ids: {string.Join(", ", duplicateTags)}").ToString());
+            }
+            var pointsJson = points.Values.Select(point => point.ToJson(tags));
+            var replacementTags = replacementTagsList.ToDictionary(tag => tag.Id, tag => tag);
+            try
+            {
+                pointsJson.Select(p => Point.FromJson(p, replacementTags));
```
Comment (Emettant) [inline server/Controllers/ApiController.cs]: here is validation of points by new tags happens; however why dont we generalize this validation to the place where we read tags: the file for tags storage does have tags, but we have no place if they are compatible with storage points

Comment (Emettant) [inline server/Controllers/ApiController.cs]: BUG: pointsJson.Select(p => Point.FromJson(p, replacementTags)) is optimized to not going inside FromJson, so I have overwritten tags couple times (that's why I have question above:] ) Thanks for having *.bak for tags in the directory.

File: server/Controllers/ApiController.cs [hunk 2]
```diff
@@ -146,6 +147,48 @@ public IActionResult PutOne(Guid id, [FromBody]PointJson pointJson)
...
+        [HttpGet("/api/tags")]
+        public IActionResult GetAllTags()
+        {
+            var tags = _memoryCache.Get<Dictionary<string, Tag>>(_tagsKey);
+            return Json(tags.Values);
+        }
+
+        [HttpPut("/api/tags")]
+        [ValidateModelState]
+        public IActionResult PutAllTags([FromBody]List<Tag> replacementTagsList)
```
Comment (Emettant) [inline server/Controllers/ApiController.cs]: personally I would dissallow to have empty [] of tags, it does not make sense.

File: server/Controllers/ApiController.cs [hunk 3]
```diff
@@ -146,6 +147,48 @@ public IActionResult PutOne(Guid id, [FromBody]PointJson pointJson)
...
+                .GroupBy(id => id)
+                .Where(g => g.Count() > 1);
+            if (duplicateTags.Any())
+            {
+                return new BadRequestObjectResult(
+                    new InvalidOperationException(
+                        $"Duplicate tag ids: {string.Join(", ", duplicateTags)}").ToString());
+            }
+            var pointsJson = points.Values.Select(point => point.ToJson(tags));
+            var replacementTags = replacementTagsList.ToDictionary(tag => tag.Id, tag => tag);
```
Comment (Emettant) [inline server/Controllers/ApiController.cs]: Dictionary is so weird: tag.Id can be "". Technically it is correct of course :) 
P.S. as well as tag.fields.name
P.S.2. Thanks God we cant have tag.fields.type to be empty :)

File: README.md [hunk 4]
```diff
@@ -54,4 +54,6 @@ Examples of the points (rename `tags.json.example` to `tags.json` for successful
 ## Methods
 * POST `/api/points` with point in the body adds a new point to the storage, validating it against corresponding tag
 * GET `/api/points` returns the entire storage contents
-* PUT `/api/points` overwrites the entire storage, the body format is the same as result of the previous method; all points are validated against tags, if any single one doesn't pass validation, storage is not modified
\ No newline at end of file
+* PUT `/api/points` overwrites the entire storage, the body format is the same as result of the previous method; all points are validated against tags, if any given one doesn't pass validation, storage is not modified
```
Comment (demyanenko) [inline README.md]: wontfix

File: server/Controllers/ApiController.cs (hunk 1 above)
Comment (demyanenko) [inline server/Controllers/ApiController.cs]: We do validate the points against tags when reading them from storage. Unifying the tags list processing makes sense, but I wouldn't do it now, as the logic is fairly simple.

File: server/Controllers/ApiController.cs [hunk 5]
```diff
@@ -146,6 +147,48 @@ public IActionResult PutOne(Guid id, [FromBody]PointJson pointJson)
...
+            try
+            {
+                pointsJson.Select(p => Point.FromJson(p, replacementTags));
+            }
+            catch(PointValidationException ex)
+            {
+                return new BadRequestObjectResult(ex.ToString());
```
Comment (demyanenko) [inline server/Controllers/ApiController.cs]: The name of the tag is included in the exception message, which is enough to pinpoint the error.

File: server/Controllers/ApiController.cs [hunk 6]
```diff
@@ -208,5 +251,19 @@ public IActionResult PutOne(Guid id, [FromBody]PointJson pointJson)
             }
             return tagsList.ToDictionary(tag => tag.Id, tag => tag);
         }
+
+        private static void WriteTags(Dictionary<string, Tag> tags)
+        {
+            if (System.IO.File.Exists(_tagsBackupPath))
```
Comment (demyanenko) [inline server/Controllers/ApiController.cs]: What exactly is duplicated?

**Pull request merged**